pkg/frontend/v2/frontend_test.go

```diff
@@ -332,8 +339,82 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	require.Nil(t, resp)
 
 	ms.checkWithLock(func() {
-		require.Equal(t, 1, len(ms.msgs))
+		require.Len(t, ms.msgs, 1)
+	})
+}
+
+// Test that the front-end shuts down gracefully, i.e. waits for in-flight queries to finish.
+func TestFrontend_GracefulShutdown(t *testing.T) {
```

Comment:
Here is alternative version of the test: it's sending more requests, and after each request is handled we check that service is still not finished, until all requests are handled.

```
func TestFrontend_GracefulShutdown(t *testing.T) {
	const (
		userID   = "test"
		body     = "all good"
		requests = 10
	)
	ctx := user.InjectOrgID(context.Background(), userID)

	inProgress := make(chan uint64, requests)
	f, _ := setupFrontend(t, nil, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
		// Remember which queries are in progress.
		inProgress <- msg.QueryID
		return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.OK}
	})

	var wg sync.WaitGroup
	for i := 0; i < requests; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			resp, err := f.RoundTripGRPC(ctx, &httpgrpc.HTTPRequest{})
			require.NoError(t, err)
			require.Equal(t, int32(200), resp.Code)
			require.Equal(t, []byte(body), resp.Body)
		}()
	}

	// Wait until all requests were sent
	test.Poll(t, 1*time.Second, requests, func() interface{} {
		return len(inProgress)
	})

	// Ask service to stop.
	f.StopAsync()

	// Service should switch to Stopping state shortly.
	test.Poll(t, 1*time.Second, services.Stopping, func() interface{} {
		return f.State()
	})

	// Verify it's not stopped yet.
	for i := 0; i < requests; i++ {
		require.Equal(t, services.Stopping, f.State())

		qid := <-inProgress

		_, err := f.QueryResult(ctx, &frontendv2pb.QueryResultRequest{
			QueryID:      qid,
			HttpResponse: &httpgrpc.HTTPResponse{Code: 200, Body: []byte(body)},
			Stats:        &stats.Stats{},
		})
		require.NoError(t, err)
	}

	// Wait until all requests handled the responses.
	wg.Wait()

	// Now the service should switch to Terminated state.
	test.Poll(t, 1*time.Second, services.Stopping, func() interface{} {
		return f.State()
	})
}
```

Reply:
I've pushed a revision with your revision, plus the extra sleep that we discussed in Slack.